_posts/2020-02-13-spring-boot-application-events-explained.md

```diff
@@ -0,0 +1,263 @@
+
+The transaction module implements an `EventListenerFactory` that looks for the new `@TransactionalEventListener` annotation. So when you annotate your method with `@TransactionalEventListener` an extended event listener that is aware of the transaction is registered instead of the default.
+
+You can bind the listener to the following phases of transaction:  
+`AFTER_COMMIT` : Fire the event after the commit has completed successfully.  
+`AFTER_COMPLETION`: Fire the event after the transaction has completed.  
+`AFTER_ROLLBACK`: Fire the event if the transaction has rolled back.  
+`BEFORE_COMMIT`: Fire the event before transaction commit.
+
+```java
+@Component
+class UserRemovedListener {
+
+  	@TransactionalEventListener(condition = "#event.name eq 'reflectoring'", phase=TransactionPhase.AFTER_COMPLETION)
+	void handleAfterUserRemoved(UserRemovedEvent event) {
+		System.out.println(String.format("User removed (@TransactionalEventListener): %s", event.getName()));
+	}
+}
+```
+
+Listener in above example will be called when `UserRemovedEvent` completes its transaction (i.e., `phase=TransactionPhase.AFTER_COMPLETION`).
+
+## Spring Boot’s Application Events
+
+Spring Boot provides a number of predefined `ApplicationEvent`s that are tied to the lifecycle of a `SpringApplication`.
+
+Some events are actually triggered before the `ApplicationContext` is created, so we cannot register a listener on those as a `@Bean`. We can register listeners for these events by adding the listener manually:
+
+```java
+@SpringBootApplication
+public class EventsDemoApplication {
+
+	public static void main(String[] args) {
+		SpringApplication springApplication = new SpringApplication(EventsDemoApplication.class);
+		springApplication.addListeners(new SpringBuiltInEventsListener());
+		springApplication.run(args);
+	}
+
+}
+```
+
+We can also register our listeners regardless of how the application is created by adding a `META-INF/spring.factories` file to our project and reference our listener(s) by using the `org.springframework.context.ApplicationListener` key:
+
+`org.springframework.context.ApplicationListener = com.reflectoring.eventdemo.SpringBuiltInEventsListener`
+
+Once we make sure that our event listener is registered properly, we can listen to all of Spring Boot's `SpringApplicationEvents`:
+
+```java
+class SpringBuiltInEventsListener implements ApplicationListener<SpringApplicationEvent>{
+
+	@Override
+	public void onApplicationEvent(SpringApplicationEvent event) {
+		System.out.println("SpringApplicationEvent Received - " + event);
+	}
+}
+```
+
+Below are the list of `SpringApplicationEvent`'s in the order of their execution,
+
+### ApplicationContextInitializedEvent
+
+An `ApplicationContextInitializedEvent` is sent when the `ApplicationContext` is prepared and `ApplicationContextInitializers` have been called but before any bean definitions are loaded.
```

Comment:
I just noticed now that the descriptions are also just copied from the docs. Please think of a use case for each event and add a sentence explaining that use case. This will increase the value by a lot.